### PR TITLE
Refactor of runtime management control flow

### LINKED ISSUE #1610: Improve RuntimeManagerMain control flow
Since https://github.com/twitter/heron/pull/1571 is getting too large, will improve `RuntimeManagerMain` in a separate PR later.

related to #1571, #1462, #1568

### PR DESCRIPTION
**Pull request by objmagic**
Similar to #1571, this PR refactors runtime management control flow.

resolves #1610

### CODE REVIEW

File: heron/spi/src/java/com/twitter/heron/spi/utils/TMasterUtils.java [hunk 1]
```diff
@@ -114,49 +114,51 @@ private static boolean sendGetRequest(URL endpoint, String command,
...
       String topologyName,
-      SchedulerStateManagerAdaptor statemgr) {
+      SchedulerStateManagerAdaptor statemgr) throws TMasterException {
     PhysicalPlans.PhysicalPlan plan = statemgr.getPhysicalPlan(topologyName);
 
     if (plan == null) {
-      LOG.log(Level.SEVERE, "Failed to get physical plan for topology {0}", topologyName);
-      return null;
+      String errMsg = String.format("Failed to get physical plan for topology '%s'", topologyName);
+      throw new TMasterException(errMsg);
```
Comment (mycFelix) [inline heron/spi/src/java/com/twitter/heron/spi/utils/TMasterUtils.java]: Wondering if we could use `Preconditions.checkState` throwing `IllegalStateException` instead of `TMasterException `

File: heron/spi/src/java/com/twitter/heron/spi/utils/TMasterUtils.java [hunk 2]
```diff
@@ -114,49 +114,51 @@ private static boolean sendGetRequest(URL endpoint, String command,
...
     }
 
     if (state == expectedState) {
-      LOG.log(Level.SEVERE, "Topology {0} command received topology {1} but already in {2} state",
-          new Object[] {topologyStateControlCommand, topologyName, state});
-      return true;
+      String errMsg = String.format(
+          "Topology {0} command received topology {1} but already in {2} state",
+          topologyStateControlCommand, topologyName, state);
+      throw new TMasterException(errMsg);
```
Comment (mycFelix) [inline heron/spi/src/java/com/twitter/heron/spi/utils/TMasterUtils.java]: using `Preconditions.checkState`?

Comment (billonahill) [inline heron/spi/src/java/com/twitter/heron/spi/utils/TMasterUtils.java]: Yes, by all means don't be shy about reviewing any PR @mycFelix.

Comment (objmagic) [inline heron/spi/src/java/com/twitter/heron/spi/utils/TMasterUtils.java]: I am not sure about using `Preconditions.checkState`. I don't see any particular benefit we could have. Besides, if we use it here, does that mean we should use it elsewhere? Does that mean we should use Guava as much as possible?

File: heron/spi/src/java/com/twitter/heron/spi/utils/TMasterUtils.java [hunk 3]
```diff
@@ -114,49 +114,51 @@ private static boolean sendGetRequest(URL endpoint, String command,
...
-      return false;
+      throw new TMasterException(String.format(
+          "Topology '%s' is not initialized yet", topologyName));
     }
 
     if (state == expectedState) {
-      LOG.log(Level.SEVERE, "Topology {0} command received topology {1} but already in {2} state",
-          new Object[] {topologyStateControlCommand, topologyName, state});
-      return true;
+      String errMsg = String.format(
```
Comment (billonahill) [inline heron/spi/src/java/com/twitter/heron/spi/utils/TMasterUtils.java]: No need to define `String errMsg` here and elsewhere

File: heron/spi/src/java/com/twitter/heron/spi/utils/TMasterUtils.java (hunk 2 above)
Comment (billonahill) [inline heron/spi/src/java/com/twitter/heron/spi/utils/TMasterUtils.java]: I favor using Preconditions.checkState for error states that should never really occur except due to a bug (i.e. dividing by 0 for example). For situations like this one though, this can actually happen if two people are trying to run a command at the same time. In this cass I think it's good to have a better typed exception.

File: heron/scheduler-core/src/java/com/twitter/heron/scheduler/TopologyRuntimeManagementException.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.twitter.heron.scheduler;
+
+/**
+ * Thrown to indicate that an error occurred while managing topology runtime
+ */
+public class TopologyRuntimeManagementException extends RuntimeException {
```
Comment (objmagic) [inline heron/scheduler-core/src/java/com/twitter/heron/scheduler/TopologyRuntimeManagementException.java]: then accordingly, `TopologySubmissionException` should be `SubmissionException`?

Comment (billonahill) [inline heron/scheduler-core/src/java/com/twitter/heron/scheduler/TopologyRuntimeManagementException.java]: Fair point. Let's keep `TopologyRuntimeManagementException`.

File: heron/scheduler-core/src/java/com/twitter/heron/scheduler/UpdateTopologyManager.java [hunk 5]
```diff
@@ -192,12 +193,9 @@ void deactivateTopology(SchedulerStateManagerAdaptor stateManager,
     logInfo("Deactivating topology %s before handling update request", topology.getName());
     NetworkUtils.TunnelConfig tunnelConfig =
         NetworkUtils.TunnelConfig.build(config, NetworkUtils.HeronSystem.SCHEDULER);
-    Preconditions.checkState(TMasterUtils.transitionTopologyState(
-        topology.getName(), TMasterUtils.TMasterCommand.DEACTIVATE, stateManager,
-        TopologyAPI.TopologyState.RUNNING, TopologyAPI.TopologyState.PAUSED, tunnelConfig),
-        String.format("Failed to deactivate topology %s. Aborting update request",
-            topology.getName()));
-
+    TMasterUtils.transitionTopologyState(
```
Comment (objmagic) [inline heron/scheduler-core/src/java/com/twitter/heron/scheduler/UpdateTopologyManager.java]: yes

File: heron/scheduler-core/src/java/com/twitter/heron/scheduler/client/SchedulerClientFactory.java [hunk 6]
```diff
@@ -55,8 +56,7 @@ public ISchedulerClient getSchedulerClient() {
           statemgr.getSchedulerLocation(Runtime.topologyName(runtime));
 
       if (schedulerLocation == null) {
-        LOG.log(Level.SEVERE, "Failed to get scheduler location");
-        return null;
+        throw new SchedulerException("Failed to get scheduler location");
```
Comment (objmagic) [inline heron/scheduler-core/src/java/com/twitter/heron/scheduler/client/SchedulerClientFactory.java]: resolved

File: heron/scheduler-core/src/java/com/twitter/heron/scheduler/utils/LauncherUtils.java [hunk 7]
```diff
@@ -109,15 +109,17 @@ public boolean onScheduleAsLibrary(
...
+      throws SchedulerException {
     String schedulerClass = Context.schedulerClass(config);
     IScheduler scheduler;
     try {
       // create an instance of scheduler
       scheduler = ReflectionUtils.newInstance(schedulerClass);
     } catch (IllegalAccessException | InstantiationException | ClassNotFoundException e) {
-      LOG.log(Level.SEVERE, "Failed to instantiate scheduler", e);
-      return null;
+      String errMsg = String.format("Failed to instantiate scheduler using class '%s'",
```
Comment (objmagic) [inline heron/scheduler-core/src/java/com/twitter/heron/scheduler/utils/LauncherUtils.java]: resolved

File: heron/spi/src/java/com/twitter/heron/spi/utils/TMasterUtils.java (hunk 3 above)
Comment (objmagic) [inline heron/spi/src/java/com/twitter/heron/spi/utils/TMasterUtils.java]: resolved

File: heron/scheduler-core/src/java/com/twitter/heron/scheduler/RuntimeManagerMain.java [hunk 8]
```diff
@@ -263,13 +266,28 @@ public static void main(String[] args)
...
-      LOG.log(Level.FINE, "Topology {0} {1} successfully", new Object[]{topologyName, command});
+    try {
+      runtimeManagerMain.manageTopology();
+      // SUPPRESS CHECKSTYLE IllegalCatch
+    } catch (Exception e) {
+      /* Since only stderr is used (by logging), we use stdout here to
+         propagate error message back to Python's executor.py (invoke site). */
+            /* Since only stderr is used (by logging), we use stdout here to
+         propagate error message back to Python's executor.py (invoke site). */
+      LOG.log(Level.FINE, "Exception when submitting topology", e);
```
Comment (objmagic) [inline heron/scheduler-core/src/java/com/twitter/heron/scheduler/RuntimeManagerMain.java]: will do this in another PR

File: heron/scheduler-core/src/java/com/twitter/heron/scheduler/RuntimeManagerMain.java [hunk 9]
```diff
@@ -263,13 +266,28 @@ public static void main(String[] args)
...
-    // Log the result and exit
-    if (!isSuccessful) {
-      throw new RuntimeException(String.format("Failed to %s topology %s", command, topologyName));
-    } else {
-      LOG.log(Level.FINE, "Topology {0} {1} successfully", new Object[]{topologyName, command});
+    try {
+      runtimeManagerMain.manageTopology();
+      // SUPPRESS CHECKSTYLE IllegalCatch
+    } catch (Exception e) {
+      /* Since only stderr is used (by logging), we use stdout here to
```
Comment (objmagic) [inline heron/scheduler-core/src/java/com/twitter/heron/scheduler/RuntimeManagerMain.java]: will do this in another PR

**Comment by billonahill** [comment]
Will review when you push your changes.

**Comment by objmagic** [comment]
reviews addressed via https://github.com/twitter/heron/pull/1618/commits/1cf64a5ce2019f577ebf341dfd2f6c0e34e588ba

File: heron/spi/src/java/com/twitter/heron/spi/utils/TMasterUtils.java [hunk 10]
```diff
@@ -65,48 +65,42 @@ public static boolean sendToTMaster(String command, String topologyName,
     try {
       URL endpoint = new URL(url);
       LOG.fine("HTTP URL for TMaster: " + endpoint);
-
-      return sendGetRequest(endpoint, command, tunnelConfig);
+      sendGetRequest(endpoint, command, tunnelConfig);
     } catch (MalformedURLException e) {
-      throw new RuntimeException("Invalid URL for tmaster endpoint: " + url, e);
+      throw new RuntimeException("Invalid URL for Tmaster endpoint: " + url, e);
```
Comment (billonahill) [inline heron/spi/src/java/com/twitter/heron/spi/utils/TMasterUtils.java]: Can we use TMasterException instead of RuntimeException? Also the TM in TMaster should both be in caps.

File: heron/spi/src/java/com/twitter/heron/spi/utils/TMasterUtils.java [hunk 11]
```diff
@@ -43,20 +44,19 @@ private TMasterUtils() {
...
+      throws TMasterException {
     // fetch the TMasterLocation for the topology
     LOG.fine("Fetching TMaster location for topology: " + topologyName);
 
     TopologyMaster.TMasterLocation location = stateManager.getTMasterLocation(topologyName);
-    if (location == null) {
-      LOG.severe("Failed to fetch TMaster Location for topology: " + topologyName);
-      return false;
-    }
+    Preconditions.checkNotNull(location, "Failed to fetch TMaster Location for topology: "
```
Comment (billonahill) [inline heron/spi/src/java/com/twitter/heron/spi/utils/TMasterUtils.java]: Sometimes we throw TMasterException and others we throw the exception generated from Proconditions. Should we be consistent and just throw TMasterException? That seems like a better experience.

Comment (objmagic) [inline heron/spi/src/java/com/twitter/heron/spi/utils/TMasterUtils.java]: consistency is nice, but does that mean we should put statement `Preconditions.checkNotNull...` inside of a `try...catch` clause. If exception `NullPointerException` gets thrown out, we catch it and throw a `TMasterException` out instead? (and [none of the `checkNotNull` accepts exception in arguments](http://google.github.io/guava/releases/snapshot/api/docs/com/google/common/base/Preconditions.html#checkNotNull(T))]

File: heron/spi/src/java/com/twitter/heron/spi/utils/TMasterUtils.java (hunk 10 above)
Comment (objmagic) [inline heron/spi/src/java/com/twitter/heron/spi/utils/TMasterUtils.java]: resolved via https://github.com/twitter/heron/pull/1618/commits/9379ed9cfdca138f46c635e3e836a81f260bcf1b

**Comment by billonahill** [comment]
👍
